src/main.py

```diff
@@ -14,28 +27,54 @@ def is_controller_connected():
 
 
 """
-Script to determine leader and follower roles.
+This function calls the ./bt_toggle script with the specified
+command and address
+"""
+def bt_toggle(command, address):
+    subprocess.call(['./bt_toggle.sh', command, address])
 
-Takes input from the camera to determine if the current vehicle
-should lead or follow based on whether or not there exists an ARTag
-in front of the vehicle.
 
-If an ARTag exists, the vehicle will become a follower vehicle.
-If an ARTag does not exist, the vehicle will become the leader vehicle.
-"""
 def main():
     tag = TagRecognition()
+    bt_address = '5C:BA:37:26:6D:9A'
 
-    # If an ARTag is detected, there is a vehicle ahead of this vehicle so it will follow.
-    # If an ARTag is not detected, no vehicle is present so this vehicle will be a leader.
-    if tag.detect() is not None:
-        follower = Follower()
-        while True:
-            follower.follow()
+    # If an ARTag is detected, the vehicle will try to disconnect from a bluetooth controller and
+    # if it is not connected to one it will follow. If an ARTag is not detected, the vehicle will
+    # will try to connect to the bluetooth controller and if succeful, the vehicle will be a leader.
+    if tag.detect():
+        bt_toggle('disconnect', bt_address)
     else:
-        leader = Leader()
-        while True:
-            leader.lead()
+        bt_toggle('connect', bt_address)
+    if is_controller_connected():
+        vehicle = Leader()
+    else:
+        vehicle = Follower()
+
+    # This loop makes vehicle move. If the vehicle is a leader and sees an ARTag
+    # it will attempt to disconnect from the bluetooth controller, if done
+    # successfully it will become a follower vehicle. If the vehicle is a follower
+    # and does not see an ARTag for more than 5 seconds, the it will try to connect
+    # to the bluetooth controller. If done successfully it will become a leader, otherwise
+    # it will continue as a follower and reset the 5 second timer before trying to
+    # connect to the bluetooth controller again.
+    start_time = time.time()
+    while True:
+        if isinstance(vehicle, Leader):
+            vehicle.lead()
+            if tag.detect():
+                bt_toggle('disconnect', bt_address)
+                if not is_controller_connected():
+                    vehicle = Follower()
+        else:
+            vehicle.follow()
+            if tag.detect():
+                start_time = time.time()
+            if (time.time() - start_time) > 5:
+                bt_toggle('connect', bt_address)     
+                if is_controller_connected():
```

Comment:
Again, Bluetooth probably won't be this quick, so I'm pretty sure 99% of the time, it'll go into this else condition. The else condition is going to reset the timer and exit the top level if, trying to look for a tag again. Then it'll just cycle

Reply:
Have you tested to make sure? Does the connection line not pause the rest of the system? Also I can't think of a fix for this without adding a sleep for 1 second to give the system a chance to connect/disconnect successfully

Reply:
I haven't tested it because I know it won't work that fast. The connection line will not pause the rest of the system. It'll send the command and continue execution. I'm not sure of a better way either. Let me think about it.